src/engine/bank05.asm

```diff
@@ -320,4 +1514,3421 @@ ZeroData: ; 1575e (5:575e)
+	ld [hli], a
+	cp $ff
+	jr nz, .loop_1
+	jp .retreat
+
+; if cost > 0 and number of energy cards attached > cost
+; choose energy cards to discard according to color
+.choose_energy_discard
+	ld a, DUELVARS_ARENA_CARD
+	call GetTurnDuelistVariable
+	call GetCardIDFromDeckIndex
+	ld a, e
+	ld [wTempCardID], a
+	call LoadCardDataToBuffer1_FromCardID
+	ld a, [wLoadedCard1Type]
+	or TYPE_ENERGY
+	ld [wTempCardType], a
+	ld a, [wTempCardRetreatCost]
+	ld c, a
+
+; first, look for and discard double colorless energy
+; if retreat cost is >= 2
+	ld hl, wDuelTempList
+	ld de, hTempRetreatCostCards
+.loop_2
+	ld a, c
+	cp 2
+	jr c, .energy_not_same_color
+	ld a, [hli]
+	cp $ff
+	jr z, .energy_not_same_color
+	ld [de], a
+	push de
+	call GetCardIDFromDeckIndex
+	ld a, e
+	pop de
+	cp DOUBLE_COLORLESS_ENERGY
+	jr nz, .loop_2
+	ld a, [de]
+	call RemoveCardFromDuelTempList
+	dec hl
+	inc de
+	dec c
+	dec c
+	jr nz, .loop_2
+	jr .end_retreat_list
+
+; second, shuffle attached cards and discard energy cards
+; that are not of the same type as the Pokémon
+; the exception for this are cards that are needed for 
+; some attacks but are not of the same color as the Pokémon
+; (i.e. Psyduck's Headache attack)
+; and energy cards attached to Eevee corresponding to a
+; color of any of its evolutions (water, fire, lightning)
+.energy_not_same_color
+	ld hl, wDuelTempList
+	call CountCardsInDuelTempList
+	call ShuffleCards
+.loop_3
+	ld a, [hli]
+	cp $ff
+	jr z, .any_energy
+	ld [de], a
+	call CheckIfEnergyIsUseful
+	jr c, .loop_3
+	ld a, [de]
+	call RemoveCardFromDuelTempList
+	dec hl
+	inc de
+	dec c
+	jr nz, .loop_3
+	jr .end_retreat_list
+
+; third, discard any card until
+; cost requirement is met
+.any_energy
+	ld hl, wDuelTempList
+.loop_4
+	ld a, [hli]
+	cp $ff
+	jr z, .set_carry
+	ld [de], a
+	inc de
+	push de
+	call GetCardIDFromDeckIndex
+	ld a, e
+	pop de
+	cp DOUBLE_COLORLESS_ENERGY
+	jr nz, .not_double_colorless
+	dec c
+	jr z, .end_retreat_list
+.not_double_colorless
+	dec c
+	jr nz, .loop_4
+
+.end_retreat_list
+	ld a, $ff
+	ld [de], a
+
+.retreat
+	ld a, $04 ; OppAction_AttemptRetreat
+	bank1call AIMakeDecision
+	or a
+	ret
+.set_carry
+	scf
+	ret
+
+; handle Mysterious Fossil and Clefairy Doll
+; if there are bench Pokémon, use effect to discard card
+; this is equivalent to using its Pokémon Power
+.mysterious_fossil_or_clefairy_doll:
+	ld a, DUELVARS_NUMBER_OF_POKEMON_IN_PLAY_AREA
+	call GetTurnDuelistVariable
+	cp 2
+	jr nc, .has_bench
+	; doesn't have any bench
+	pop af
+	jr .set_carry
+
+.has_bench
+	ld a, DUELVARS_ARENA_CARD
+	call GetTurnDuelistVariable
+	ldh [hTempCardIndex_ff9f], a
+	xor a
+	ldh [hTemp_ffa0], a
+	ld a, $0c ; OppAction_UsePokemonPower
+	bank1call AIMakeDecision
+	pop af
+	ldh [hTempPlayAreaLocation_ffa1], a
+	ld a, $0d ; OppAction_ExecutePokemonPowerEffect
+	bank1call AIMakeDecision
+	ld a, $16 ; OppAction_DrawDuelMainScene
+	bank1call AIMakeDecision
+	or a
+	ret
+; 0x15ea6
+
+; Copy cards from wDuelTempList to wHandTempList
```

Comment:
I would change to "Copy cards from wDuelTempList in hl to wHandTempList in de" (the name is fine since the function is only used for that at it would've been a local function to ``Func_15eae`` if it wasn't because ``Func_15f4c``, but I think the clarification is helpful.